{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5421844f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "plt.style.use('seaborn-v0_8-darkgrid')\n",
    "sns.set_palette(\"husl\")\n",
    "from matplotlib.ticker import FuncFormatter\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "e4459f57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Общий объем данных: 2.7105 TB\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('SraRunTable.csv', index_col=False)\n",
    "total_bytes = df['Bytes'].sum()\n",
    "print(f\"Общий объем данных: {total_bytes/1e12:.4f} TB\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d40e82f",
   "metadata": {},
   "source": [
    "### Чистим от 16S rRNA было 2807 - стало 1707"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "90c44bcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "deleted_rows = df[df['target_gene (exp)'] == '16S rRNA']\n",
    "df = df[df['target_gene (exp)'] != '16S rRNA']\n",
    "#df\n",
    "#deleted_rows"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc9a8c11",
   "metadata": {},
   "source": [
    "### Убираем запуски с аномально большим количеством баз (больше 30 гигабаз и весят по 50-150 гигабайт) (пока не убираем)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "a00c1bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df = df[df['Bases'].apply(lambda x: x<30000000000)]\n",
    "#df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28a2c5a7",
   "metadata": {},
   "source": [
    "### Удаляем хозяев мышь и выдру"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "05122107",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[(df['HOST'] != 'Mus musculus') & (df['HOST'] !='southern_sea_otter')]\n",
    "#df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93bf1073",
   "metadata": {},
   "source": [
    "### Удаляем проект PRJNA1214429 с 311 архивами, в которых все прочтения бактериальные и PRJNA665586 (4 архива) просто мазок из пещеры - вирусных прочтений нет."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "e13b2814",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[(df['BioProject'] != 'PRJNA1214429') & (df['BioProject'] != 'PRJNA665586')]\n",
    "#df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dc33d34",
   "metadata": {},
   "source": [
    "### Удаляем проекты, где организм бактерии"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53a249bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Удалено 130 строк:\n"
     ]
    }
   ],
   "source": [
    "\n",
    "pattern = r'(Escherichia coli|Lactobacillus|Lactococcus sp\\.|Escherichia marmotae|Clostridium sp\\.|Enterobacteriaceae bacterium|Streptomyces sp\\.|Staphylococcus nepalensis)'\n",
    "\n",
    "mask = ~df['Organism'].str.contains(pattern, case=False, na=False, regex=True)\n",
    "filtered_df = df[mask]\n",
    "\n",
    "removed = df[~mask]\n",
    "print(f\"Удалено {len(removed)} строк:\")\n",
    "#print(removed['Organism'].unique())\n",
    "df = filtered_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f913ac3d",
   "metadata": {},
   "source": [
    "### Убираем пустые столбцы. В итоге остается 1189 архивов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "2e9e87c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.dropna(axis=1, how='all')\n",
    "#df.to_csv('sra_filtered.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "91fc5ee5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df['tissue'].value_counts()\n",
    "#df['HOST'].value_counts()\n",
    "#df['Organism'].value_counts()\n",
    "#df.loc[df['BioProject'] == 'PRJNA665586']\n",
    "#df.loc[df['HOST'] == 'Environment']\n",
    "#df.loc[df['isolation_source'] == 'forest']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dc9353b",
   "metadata": {},
   "source": [
    "### Данные весят 2.58 терабайт"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "9038d111",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Общий объем данных: 2.5857 TB\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('sra_filtered.csv')\n",
    "total_bytes = df['Bytes'].sum()\n",
    "print(f\"Общий объем данных: {total_bytes/1e12:.4f} TB\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "e9006ed9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Большие файлы:\n",
      "             Run     Size_GB   Size_TB Assay Type            Organism\n",
      "733  SRR23314745  161.519252  0.161519    RNA-Seq      bat metagenome\n",
      "736  SRR23314746  123.813281  0.123813    RNA-Seq      bat metagenome\n",
      "735  SRR23314743  100.071383  0.100071    RNA-Seq      bat metagenome\n",
      "737  SRR23314747   61.614715  0.061615    RNA-Seq      bat metagenome\n",
      "740  SRR23314751   57.810831  0.057811    RNA-Seq      bat metagenome\n",
      "739  SRR23314750   53.888778  0.053889    RNA-Seq      bat metagenome\n",
      "198  SRR19599582   34.619077  0.034619    RNA-Seq  Eonycteris spelaea\n",
      "738  SRR23314749   34.270534  0.034271    RNA-Seq      bat metagenome\n",
      "429  SRR14860617   33.661649  0.033662        WGS      gut metagenome\n",
      "197  SRR19599581   33.630739  0.033631    RNA-Seq  Eonycteris spelaea\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nБольшие файлы:\")\n",
    "top_large = df.nlargest(10, 'Bytes')[['Run', 'Bytes', 'Assay Type', 'Organism']].copy()\n",
    "top_large['Size_GB'] = top_large['Bytes'] / 1e9\n",
    "top_large['Size_TB'] = top_large['Bytes'] / 1e12\n",
    "print(top_large[['Run', 'Size_GB', 'Size_TB', 'Assay Type', 'Organism']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "095f89c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Assay Type:\n",
      "  Количество уникальных значений: 3\n",
      "  Топ-10 значений: {'RNA-Seq': 576, 'WGS': 498, 'OTHER': 115}\n",
      "\n",
      "LibraryLayout:\n",
      "  Количество уникальных значений: 2\n",
      "  Топ-10 значений: {'PAIRED': 1100, 'SINGLE': 89}\n",
      "\n",
      "LibrarySource:\n",
      "  Количество уникальных значений: 7\n",
      "  Топ-10 значений: {'METAGENOMIC': 797, 'METATRANSCRIPTOMIC': 186, 'GENOMIC': 108, 'VIRAL RNA': 47, 'TRANSCRIPTOMIC': 43, 'TRANSCRIPTOMIC SINGLE CELL': 4, 'OTHER': 4}\n",
      "\n",
      "Platform:\n",
      "  Количество уникальных значений: 7\n",
      "  Топ-10 значений: {'ILLUMINA': 1166, 'OXFORD_NANOPORE': 7, 'DNBSEQ': 6, 'PACBIO_SMRT': 5, 'CAPILLARY': 3, 'ION_TORRENT': 1, 'GENEMIND': 1}\n",
      "\n",
      "geo_loc_name_country:\n",
      "  Количество уникальных значений: 38\n",
      "  Топ-10 значений: {'Ghana': 221, 'China': 171, 'Panama': 101, 'Spain': 99, 'Switzerland': 70, 'Singapore': 62, 'USA': 43, 'uncalculated': 41, 'Australia': 38, 'South Korea': 31}\n",
      "\n",
      "geo_loc_name_country_continent:\n",
      "  Количество уникальных значений: 7\n",
      "  Топ-10 значений: {'Asia': 308, 'Europe': 268, 'Africa': 261, 'North America': 161, 'uncalculated': 41, 'Oceania': 38, 'South America': 3}\n",
      "\n",
      "sample_type:\n",
      "  Количество уникальных значений: 2\n",
      "  Топ-10 значений: {'Metagenomic assembly': 12, 'bat feces': 1}\n"
     ]
    }
   ],
   "source": [
    "categorical_cols = ['Assay Type', 'LibraryLayout', 'LibrarySource', 'Platform', 'geo_loc_name_country', 'geo_loc_name_country_continent', 'sample_type']\n",
    "for col in categorical_cols:\n",
    "    if col in df.columns:\n",
    "        print(f\"\\n{col}:\")\n",
    "        print(f\"  Количество уникальных значений: {df[col].nunique()}\")\n",
    "        print(f\"  Топ-10 значений: {df[col].value_counts().head(10).to_dict()}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "virus",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
